src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,287 @@
+module Oscoin.Protocol
+    ( Protocol -- opaque
+    , Handle -- opaque
+    , ShutdownPolicy(..)
+
+    -- * Getters
+    , protoOrphanage
+
+    -- * Running and using the Protocol
+    , withProtocol
+    , runOscoinProtocol
+
+    -- * Internals
+    , stepProtocol
+    ) where
+
+import           Oscoin.Prelude
+
+import           Formatting (Buildable, Format, shown, (%))
+import           Oscoin.Consensus (validateBlockchain)
+import           Oscoin.Consensus.Config as Consensus
+import           Oscoin.Consensus.Types (Validate)
+import           Oscoin.Crypto.Blockchain (Blockchain(..))
+import           Oscoin.Crypto.Blockchain.Block
+import           Oscoin.Crypto.Hash (HasHashing, Hash)
+import           Oscoin.Storage.Block.Abstract
+                 ( BlockStore
+                 , BlockStoreInternals
+                 , FullBlockStoreAPI
+                 , GetBlocksQuery(..)
+                 )
+import qualified Oscoin.Storage.Block.Abstract as BlockStore
+import           Oscoin.Storage.Block.Orphanage (Orphanage)
+import qualified Oscoin.Storage.Block.Orphanage as O
+import           Oscoin.Telemetry (fmtBlockHash, fmtValidationError)
+import           Oscoin.Telemetry.Logging (LogMsg, ftag)
+import           Oscoin.Time.Chrono as Chrono
+
+import           Control.Concurrent.Async (Async)
+import qualified Control.Concurrent.Async as Async
+import           Control.Concurrent.STM
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | A stateful handle over a 'Protocol'.
+data Handle c tx s = Handle
+    { hProtocol :: MVar (Protocol c tx s IO) }
+
+-- | This data structures incorporates all the different components and
+-- data structures in the system devoted to store \"things\". Certain
+-- components (like the block cache, for example) are not managed by the
+-- 'Protocol' directly but they are rather an implementation detail of
+-- whichever 'BlockStore' this manager encapsulates.
+data Protocol c tx s m = Protocol
+    { protoFullBlockStore :: FullBlockStoreAPI c tx s m
+    -- ^ The 'BlockStore', an opaque interface to a component which only
+    -- responsibility is to store and retrieve blocks.
+    , protoOrphanage      :: Orphanage c tx s
+    -- ^ The in-memory data structure which keeps orphans blocks and constructs
+    -- sub-chains suitable for fork selection.
+    , protoFetchNextBlock :: m (Maybe (Block c tx (Sealed c s)))
```

Comment:
In the current implementation the action always blocks until a new block is available. I guess we can remove `Maybe`?

Reply:
Having a `Maybe` there gives us more flexibility methinks, especially in tests. It doesn't cost us much, but (for example) gives us the ability to "peek" or trying to read into/from the queue (cfr. `peekTBQueue` and `tryReadTBQueue`).